sample/App.js

```diff
@@ -1,96 +1,66 @@
-import React, {useState, useEffect} from 'react';
-import {
-  SafeAreaView,
-  Text,
-  StyleSheet,
-  Button,
-  AppState,
-  View,
-} from 'react-native';
-import RNAndroidNotificationListener from 'react-native-android-notification-listener';
+import React, { useState, useEffect } from 'react'
+import { SafeAreaView, Text, Button, AppState, View, AppRegistry } from 'react-native'
+import RNAndroidNotificationListener from 'react-native-android-notification-listener'
+import styles from './App.styles.js'
 
 const App = () => {
-  const [hasPermission, setHasPermission] = useState(false);
-  const [lastNotification, setLastNotification] = useState(null);
+    const [hasPermission, setHasPermission] = useState(false)
+    const [lastNotification, setLastNotification] = useState(null)
 
-  const handleOnPressPermissionButton = async () => {
-    RNAndroidNotificationListener.requestPermission();
-  };
+    const handleOnPressPermissionButton = async () => {
+        RNAndroidNotificationListener.requestPermission()
+    }
 
-  const handleAppStateChange = async nextAppState => {
-    RNAndroidNotificationListener.getPermissionStatus().then(status => {
-      setHasPermission(status !== 'denied');
-    });
-  };
+    const handleAppStateChange = async nextAppState => {
+        if (nextAppState === 'active') {
+            RNAndroidNotificationListener.getPermissionStatus().then(status => {
+                setHasPermission(status !== 'denied')
+            })
+        }
+    }
 
-  const handleNotificationReceived = notification => {
-    setLastNotification(notification);
-  };
+    // THIS HANDLE MUST BE ASYNCRONOUS
+    const headlessNotificationListener = async notification => {
+        if (notification) {
+            setLastNotification(notification)
+        }
+    }
 
-  useEffect(() => {
-    RNAndroidNotificationListener.getPermissionStatus().then(status => {
-      setHasPermission(status !== 'denied');
-    });
+    useEffect(() => {
+        AppState.addEventListener('change', handleAppStateChange)
 
-    RNAndroidNotificationListener.onNotificationReceived(
-      handleNotificationReceived,
-    );
+        AppRegistry.registerHeadlessTask(
```

Comment:
I don't think this is the right place to register this task, even though the effect is correctly set up to only be run once.

This should be in `sample/index.js` and use a module import as needed to register the correct function.

I guess this conflicts with the simple example though, which just uses React state without needing any sort of external store? If you can't think of an easy way to do this without refactoring the whole sample, I'd just add a comment about this right here.

From Docs: [https://reactnative.dev/docs/appregistry]
>AppRegistry should be required early in the require sequence to make sure the JS execution environment is setup before other modules are required.

Reply:
Actually now that I think about it, it's a strong possibility this will not work at all.

When RN Android boots up the JS looking for a headless task to execute, I'm not 100% sure it will actually mount a UI for the normal RN bridge to interact with. I suspect it won't, which means your React UI won't be mounted, and this effect will never be triggered. Since the task is never registered, it will likely fail with a logged error.

Reply:
Nice observation @jehartzog ... I'll change the implementation to `sample/index` and do some tests using AsyncStorage to store the notification payloads to see how the service will respond, even after restarting the device and closing the application. The [BootUpReceiver](https://github.com/leandrosimoes/react-native-android-notification-listener/pull/6/files#diff-22c310e837f24cbedd8672e9c6417b8f091cd024012686212f629a5733229204) should start the service automatically as soon as the device restart.

I'll do some tests and let you know.

Reply:
@jehartzog ... I did the changes, put the `AppRegistry` and the handle function in the `sample/index`. Inside the handle function, I put a code to store the notification in the AsyncStorage. After that, I install the release version of the sample application on my device and restarted it. As soon as my device finished to boot up, and without open the sample application, I sent a new notification to another app on my device and received it. Then I open the sample application and the notification was there in the AsyncStorage! So I think that this way will work just fine.

Any other thoughts to be considered here?

Reply:
Nope, `AsyncStorage` is the perfect solution to show how this needs to work. RN devs will understand it easily, and those using some sort of external data store/API to actually use the functionality will easily be able to modify it to work for their use case. The updated example helps show how the `headlessJS` portion is actually completely separate from the `React` UI, and so you need some external data linkage (`AsyncStorage`  or API based) to make it functional.

I didn't actually see the updates to this PR since you haven't pushed them yet, but the idea sounds good.

Reply:
Awesome @jehartzog ... Sorry about not having pushed yet, I just wanted to confirm with you before pushing but now is all pushed. Take a look and let me know if everything looks fine, pls!

Reply:
@leandrosimoes Looks solid.

Reply:
Awesome @jehartzog, thanks a lot for the help. I'll merge this branch and release a version in a few minutes.